lib/eth/abi.rb

```diff
@@ -381,6 +410,39 @@ def encode_bytes(arg, type)
       end
     end
 
+    # Properly encodes tuples.
+    def encode_tuple(arg, type)
+      raise EncodingError, "Expecting Hash: #{arg}" unless arg.instance_of? Hash
+      raise EncodingError, "Expecting #{type.components.size} elements: #{arg}" unless arg.size == type.components.size
+
+      static_size = 0
+      type.components.size.times do |i|
+        if type.components[i].is_dynamic?
+          static_size += 32
+        else
+          static_size += Util.ceil32(type.components[i].size || 0)
+        end
+      end
+
+      dynamic_offset = static_size
+      offsets_and_static_values = []
+      dynamic_values = []
+
+      type.components.size.times do |i|
+        component_type = type.components[i]
+        if component_type.is_dynamic?
+          offsets_and_static_values << encode_type(Type.size_type, dynamic_offset)
+          dynamic_value = encode_type(component_type, arg.is_a?(Array) ? arg[i] : arg[component_type.name])
+          dynamic_values << dynamic_value
+          dynamic_offset += dynamic_value.size
+        else
+          offsets_and_static_values << encode_type(component_type, arg.is_a?(Array) ? arg[i] : arg[component_type.name])
```

Comment:
Maybe make this a method:

`encode_type(component_type, arg.is_a?(Array) ? arg[i] : arg[component_type.name])` since you use it more than once